main/tests/server/src/com/google/refine/RefineTest.java

```diff
@@ -449,4 +451,58 @@ protected void runAndWait(ProcessManager processManager, Process process, int ti
     public static void assertEqualsSystemLineEnding(String actual, String expected) {
         Assert.assertEquals(actual, expected.replaceAll("\n", System.lineSeparator()));
     }
+
+    /**
+     * Checks that the grid contents are equal in both projects. Differences in "cell indices" are not taken into
+     * account as this is an internal detail: the goal is to assert equality of the user-facing parts of project data.
+     * 
+     * @param actual
+     *            the actual project state
+     * @param expected
+     *            the expected project state
+     */
+    public static void assertProjectEquals(Project actual, Project expected) {
```

Comment:
For the test cases included in this PR, it seems like the `expected` project is always something which has just been created by passing an array of column names and a grid of values to the `createProject()` method and this comparison only considers the grid, not other aspects of a `Project`, so I'm wondering if it wouldn't be simpler and more direct to just compare directly against the column names & grid. This would allow you to skip half of the cell index redirection. This would also avoid the semantic mismatch if we want to introduce a real projectEquals assertion in the future.

Are there other places where the project creation is further removed from the test assertion?

Reply:
I prefer keeping the assertion separate from the creation of the expected project, as it feels architecturally cleaner to me. That's primarily motivated by my further changes that this prepares for: introducing the `Grid` class, an immutable class which represents the state of the project at a given point in the history. Once that is done, this test utility will be adapted to compare Grid objects instead. This would let you do things such as:
* testing that an operation has no effect on a grid: just check that the original grid is equal to the one that is returned by the operation. If we were to inline grid creation inside the test utility, you'd need to supply again the test data the grid was created from.
* even without introducing `Grid`, there are sometimes aspects of the project's state that aren't so easy to set up via `createProject` and which should rather be added manually afterwards (say, defining column groups). Keeping the assertion separate from the creation makes it possible to add those details and then do the assertion.

> This would also avoid the semantic mismatch if we want to introduce a real projectEquals assertion in the future.

What do you mean by that? What would a real `projectEquals` assertion look like?

Reply:
I wasn't proposing creating a project, but rather just using the test data grid directly. It seems like String[] columnNames, Serializable[][] values is effectively your grid anyway, but I don't feel strongly enough to argue about it (and I don't know what this future state looks like).

> What would a real `projectEquals` assertion look like?

It would compare other non-grid aspects of the project like its metadata, change history, etc.

Reply:
Ah yes I see. Then I think the introduction of the `Grid` class will go in the direction you are aiming at, since that will be an object that represents just the current project state (without history, project id and other things like that).

Checking for equality on entire projects (including the history) could be useful in certain cases but I don't think it would be convenient for operation tests to rely on this, typically because history metadata contains timestamps and localized information such as descriptions that would be annoying to equate exactly.